autonav_ws/src/autonav_shared/src/node.cpp

```diff
@@ -0,0 +1,268 @@
+    }
+
+    void Node::device_state_callback(const autonav_msgs::msg::DeviceState::SharedPtr msg)
+    {
+        // Log teh raw message
+        if (msg->device == get_name())
+        {
+            log("Received update on our device state from " + AutoNav::DEVICE_STATE_NAMES.at(device_states.at(msg->device)) + " to " + AutoNav::DEVICE_STATE_NAMES.at(static_cast<AutoNav::DeviceState>(msg->state)), Logging::LogLevel::DEBUG);
+        }
+
+        if (msg->device != get_name())
+        {
+            device_states.insert_or_assign(msg->device, static_cast<AutoNav::DeviceState>(msg->state));
+            return;
+        }
+        
+        // TODO: Unravel this
+        if (device_states.find(msg->device) == device_states.end())
+        {
+            init();
+        } else {
+            if (device_states.at(msg->device) == AutoNav::DeviceState::OFF && static_cast<AutoNav::DeviceState>(msg->state) == AutoNav::DeviceState::WARMING)
+            {
+                init();
+            }
+        }
+    }
+
+    // TODO: Log to file
+    void Node::log(const std::string & message, Logging::LogLevel level, const std::string & function_caller, int line)
+    {        
+        // Current time/date
+        auto now = std::chrono::system_clock::now();
+        std::time_t now_time = std::chrono::system_clock::to_time_t(now);
+        std::tm now_tm = *std::localtime(&now_time);
+        std::chrono::milliseconds ms = std::chrono::duration_cast<std::chrono::milliseconds>(now.time_since_epoch()) % 1000;
+        
+        // Log level
+        std::string log_level = AutoNav::Logging::LOG_LEVEL_NAMES.at(level);
+        if (log_level.length() < 5)
+        {
+            log_level += std::string(5 - log_level.length(), ' ');
+        }
+
+        // Log to topic
+        autonav_msgs::msg::Log log_msg;
+        log_msg.timestamp = now_time;
+        log_msg.line_number = line;
+        log_msg.function_caller = function_caller;
+        log_msg.message = message;
+        log_msg.node = get_name();
+        log_msg.level = static_cast<uint8_t>(level);
+        log_pub->publish(log_msg);
+
+        switch (level)
+        {
+            case Logging::LogLevel::DEBUG:
+                std::cout << termcolor::color<99, 150, 79> << std::put_time(&now_tm, "%Y-%m-%d %H:%M:%S") << ":" << std::setfill('0') << std::setw(3) << ms.count() 
+                    << termcolor::white << " | " 
+                    << termcolor::color<61, 117, 157> << log_level 
+                    << termcolor::white << " | " 
+                    << termcolor::color<90, 60, 146> << get_name() 
+                    << termcolor::white << ":" 
+                    << termcolor::color<90, 60, 146> << function_caller 
+                    << termcolor::white << ":" 
+                    << termcolor::color<90, 60, 146> << line 
+                    << termcolor::white << " - " 
+                    << termcolor::color<61, 117, 157> << message << termcolor::reset << std::endl;
```

Comment:
if possible make print statements not repetitive but C++ bad so if you can't variablize this then that's fine

Reply:
not possible :(